Operational Space Control

yminchen opened the pull request:
Major updates:
- Added `OperationalSpaceControl` and the related classes
- Added standing controller for Cassie based on operational space control
- Added walking controller for Cassie based on operational space control and capture point foot placement

Minor updates:
- Added time based finite state machine
- Updated `cassie_utils` to create fixed spring cassie model
- Updated `cassie_fixed_springs.urdf` so that it's the same as `cassie_v2.urdf` except the springs
- Updated `cassie_v2.urdf` to include toe motor limit (it was infinity originally)
- Added a util function to check if a RBT is floating based
- Added flags to initialize Cassie's pose in `run_simple_sim.cc`
- Added a util function to avoid of 0-norm quaternion

mposa commented on systems/controllers/cp_traj_gen.cc:
```diff
@@ -0,0 +1,397 @@
+                BasicVector<double>(1)).get_index();
+  if (is_using_predicted_com) {
+    com_port_ = this->DeclareAbstractInputPort("CoM_traj",
+        drake::Value<ExponentialPlusPiecewisePolynomial<double>> {}).get_index();
```
Is such a specific class `ExponentialPlusPiecewisePolynomial` needed here, or could this be a more generic `Trajectory`?

yminchen replied:
I once tried to use `Trajectory`, but I couldn't find away to initialize the Abstract input port, because `Trajectory` is a virtual class.
However, it will be great if we can do so. It'll make OSC class a little bit cleaner.

mposa replied:
I see. I took a quick look, and didn't see a way around that issue.

mposa commented on systems/controllers/cp_traj_gen.cc:
```diff
@@ -0,0 +1,397 @@
+  // State variables inside this controller block
+  DeclarePerStepDiscreteUpdateEvent(&CPTrajGenerator::DiscreteVariableUpdate);
+  // The swing foot position at touchdown
+  prev_td_swing_foot_idx_ = this->DeclareDiscreteState(3);
```
Isn't this just the stance foot position?

yminchen replied:
It is both the stance foot position at the end of stance phase, and the swing foot position at the start of swing phase.

mposa replied:
Ah, that wasn't clear.

mposa commented on systems/controllers/cp_traj_gen.cc:
```diff
@@ -0,0 +1,397 @@
+    Vector3d pt_on_swing_foot = (fsm_state(0) == right_stance_) ?
+                                pt_on_left_foot_ : pt_on_right_foot_;
+
+    // Swing foot position (Forward Kinematics) and velocity at touchdown
```
This forward kinematics can be done via a single Drake call, you should almost never have to manually deal with rotation matrices. This should do the trick (where the `0` is the world index):

```
swing_foot_pos_td = tree_->transformPoints(cache, pt_on_swing_foot, swing_foot_idx, 0);
```

mposa commented on systems/controllers/cp_traj_gen.cc:
```diff
@@ -0,0 +1,397 @@
+  VectorXd fsm_state = fsm_output->get_value();
+
+  // Get stance foot position and index
+  KinematicsCache<double> cache = tree_->CreateKinematicsCache();
```
This comment goes for this file and others--you might be paying a significant computational penalty because of the way you split up these methods _without_ leveraging caching. Many of your methods recompute the same `KinematicsCache`. `MBP` will help with this, but you also might be better off by passing the cache as inputs to many methods.

yminchen replied:
What is the best way of doing the caching? I'm thinking create a class `CentralKinematicsCache` with two members
- `kinematics_cache`
- `time` which stores the time when the kinematics cache is updated.

 and two functions
- `initialize()` which initiailizes the members
- `update()` which updates kinematics cache if the current time is greater than `time`, and return a const reference to `kinematics_cache`.

I construct `CentralKinematicsCache` in run_osc_walking_controller.cc and pass it into the leaf systems that need it?

mposa commented on systems/controllers/cp_traj_gen.cc:
```diff
@@ -0,0 +1,397 @@
+  Eigen::MatrixXd stance_foot_body_rot = stance_foot_body_pose.linear();
+  Vector3d stance_foot_pos = stance_foot_body_pos +
+                             stance_foot_body_rot * pt_on_stance_foot;
+
```
Extra white space here

mposa commented on systems/controllers/cp_traj_gen.cc:
```diff
@@ -0,0 +1,397 @@
+
+  // Get CoM or predicted CoM
+  Vector3d CoM;
+  Vector3d dCoM;
```
`dCOM=velocity`?

yminchen replied:
Yes, should I change the name?

yminchen commented on examples/Cassie/run_osc_standing_controller.cc:
```diff
@@ -0,0 +1,236 @@
+// publish rate of the robot state needs to be less than 500 Hz. Otherwise, the
+// performance seems to degrade due to this. (Recommended publish rate: 200 Hz)
+// Maybe we need to update the lcm driven loop to clear the queue of lcm message
+// if it's more than one message?
```
What is the reason of doing it asynchronously?

mposa replied:
I take that back, either is fine.

Asynchronously means we can run the OSC at whatever rate is necessary, without worrying about the LCM queue (though I'm pretty sure the LCM driven loop already clears it's queue, though I could be wrong).

yminchen commented on examples/Cassie/osc_walking_control/foot_placement_control.h:
```diff
@@ -0,0 +1,60 @@
+ public:
+  FootPlacementControl(RigidBodyTree<double>* tree,
+    int pelvis_idx,
+    Eigen::Vector2d global_target_position, double circle_radius_of_no_turning);
```
Currently it's a constant just to show that the controller is working. We can extend it to an input of a block in the future if it's needed?

mposa commented:
Ok. Maybe mark that as a TODO.

On Mon, Jul 22, 2019 at 11:28 PM Yu-Ming Chen <notifications@github.com>
wrote:

> *@yminchen* commented on this pull request.
> ------------------------------
>
> In examples/Cassie/osc_walking_control/foot_placement_control.h
> <https://github.com/DAIRLab/dairlib/pull/86#discussion_r306115335>:
>
> > +namespace cassie {
> +namespace osc_walking_control {
> +
> +// TODO(yminchen): we can replace cp with raibert style control. (feedforward
> +// term is v*T/2)
> +
> +
> +/// FootPlacementControl calculates the deviation from capture point in order
> +/// to track a desired velocity.
> +
> +/// Requirement: quaternion floating-based Cassie only
> +class FootPlacementControl : public drake::systems::LeafSystem<double> {
> + public:
> +  FootPlacementControl(RigidBodyTree<double>* tree,
> +    int pelvis_idx,
> +    Eigen::Vector2d global_target_position, double circle_radius_of_no_turning);
>
> Currently it's a constant just to show that the controller is working. We
> can extend it to an input of a block in the future if it's needed?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/DAIRLab/dairlib/pull/86?email_source=notifications&email_token=ABBICGC5WISXORM5OVTWXZ3QAZ3EHA5CNFSM4IF2QYCKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB7G7HAQ#discussion_r306115335>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBICGB43DZGM6AHBLJFVI3QAZ3EHANCNFSM4IF2QYCA>
> .
>

mposa commented:
I would recommend changing it. I don't think that `dVar` is commonly used
to mean `dVar/dt`

On Mon, Jul 22, 2019 at 10:58 PM Yu-Ming Chen <notifications@github.com>
wrote:

> *@yminchen* commented on this pull request.
> ------------------------------
>
> In systems/controllers/cp_traj_gen.cc
> <https://github.com/DAIRLab/dairlib/pull/86#discussion_r306110760>:
>
> > +  } else {
> +    stance_foot_idx = left_foot_idx_;
> +    pt_on_stance_foot = pt_on_left_foot_;
> +  }
> +  Eigen::Isometry3d stance_foot_body_pose =
> +      tree_->CalcBodyPoseInWorldFrame(cache, tree_->get_body(stance_foot_idx));
> +  Vector3d stance_foot_body_pos = stance_foot_body_pose.translation();
> +  Eigen::MatrixXd stance_foot_body_rot = stance_foot_body_pose.linear();
> +  Vector3d stance_foot_pos = stance_foot_body_pos +
> +                             stance_foot_body_rot * pt_on_stance_foot;
> +
> +
> +
> +  // Get CoM or predicted CoM
> +  Vector3d CoM;
> +  Vector3d dCoM;
>
> Yes, should I change the name?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/DAIRLab/dairlib/pull/86?email_source=notifications&email_token=ABBICGAY3E3FSFEQJFLM4KDQAZXT7A5CNFSM4IF2QYCKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB7G54QQ#discussion_r306110760>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBICGH37EEPO6WX7FOXEBTQAZXT7ANCNFSM4IF2QYCA>
> .
>

yminchen commented on examples/Cassie/osc_walking_control/cp_control_common_func.cc:
```diff
@@ -0,0 +1,24 @@
+
+namespace dairlib {
+namespace cassie {
+namespace osc_walking_control {
```
I thought that we might have more osc controllers in the future, so I named it as such. maybe I can change it to `osc_walk`?

mposa commented:
More OSC's could still exist within the same `namespace`. I'll leave it up
to you.

On Mon, Jul 22, 2019 at 11:32 PM Yu-Ming Chen <notifications@github.com>
wrote:

> *@yminchen* commented on this pull request.
> ------------------------------
>
> In examples/Cassie/osc_walking_control/cp_control_common_func.cc
> <https://github.com/DAIRLab/dairlib/pull/86#discussion_r306116188>:
>
> > @@ -0,0 +1,24 @@
> +#include "examples/Cassie/osc_walking_control/cp_control_common_func.h"
> +
> +using Eigen::Vector2d;
> +
> +namespace dairlib {
> +namespace cassie {
> +namespace osc_walking_control {
>
> I thought that we might have more osc controllers in the future, so I
> named it as such. maybe I can change it to osc_walk?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/DAIRLab/dairlib/pull/86?email_source=notifications&email_token=ABBICGCPWXNDNKDYULNP6W3QAZ3VXA5CNFSM4IF2QYCKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB7G7O5A#discussion_r306116188>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBICGHERRCQ4MWZVVI2DFTQAZ3VXANCNFSM4IF2QYCA>
> .
>

yminchen commented on examples/Cassie/cassie_utils.cc:
```diff
@@ -100,30 +100,32 @@ void buildCassieTree(RigidBodyTree<double>& tree, std::string filename,
-  RevoluteJoint& ankle_spring_joint_right =
-      dynamic_cast<RevoluteJoint&>(body->get_mutable_joint());
-  ankle_spring_joint_right.SetSpringDynamics(1250.0, 0.0);  // 2300 in URDF
+  if(is_with_springs){
```
I'm not very sure how `Weld` works. Does it means that two body cannot move relatively anymore? If it is what I think, maybe we don't need `Weld`, because we still want the joints to move freely?

mposa replied:
Yes, that's how Weld works. My thought was that, rather than using a separate URDF, we could simply Weld the spring joint for the fixed-springs version. (fixed springs, not no springs)

yminchen replied:
I wonder if welding works in our case. In cassie_v2.urdf, the bodies relation is "thigh - shin" instead of "thigh - knee_spring - shin"

mposa replied:
I'm not sure why it wouldn't work. Each spring acts on a joint (which connects two bodies). Those two bodies could just be welded.

yminchen replied:
Oh I think that the knee spring is _not_ connected to both thigh and shin in `cassie_v2.urdf`?
I think it's look like: thigh ------ shin
                                       \
                                        -------knee spring
The only joint that `knee_spring` body is connected to is `knee_spring_left_fixed`.
https://github.com/DAIRLab/dairlib/blob/c6f4fc414010c03c9a0801f85adc500a1a737531/examples/Cassie/urdf/cassie_v2.urdf#L733

mposa replied:
Resolved f2f.

yminchen commented on examples/Cassie/cassie_test.pmd:
```diff
@@ -30,10 +30,22 @@ group "1.simulated-robot" {
         host = "localhost";
     }
+    cmd "1.simulator (no dispatcher) for OSC controller" {
+        exec = "bazel-bin/examples/Cassie/run_simple_sim --floating_base=true --publish_state=true --publish_rate=200";
```
I agree. I felt it's getting messy when I updated it.

yminchen commented on common/math_utils.cc:
```diff
@@ -0,0 +1,43 @@
+
+namespace dairlib {
+
+Eigen::Quaterniond NormalizeQuaternion(Eigen::Quaterniond q) {
```
Based on the discussion yesterday, I think we decided to leave it as it is for now, because we have the 0-norm quaternion case.

mposa replied:
I disagree. A method called "normalize" should never reset the quaternion. This check should be inserted explicitly into blocks that need to "fix" any 0-norm vectors, and a warning message should be issued.

yminchen replied:
Ok, I agree that the 0-norm checking shouldn't hide in a function named normalize(). I changed the function to `CheckZeroQuaternion` and moved it to `rigidbody_utils`

mposa replied:
Name suggestion: `SetZeroQuaternionToIdentity`

yminchen commented on systems/controllers/cp_traj_gen.h:
```diff
@@ -0,0 +1,135 @@
+
+  bool is_quaternion_;
+
+  RigidBodyTree<double>* tree_;
```
Done. I changed to pointer for some reason that I have to use a non-const function, but it doesn't seem to be a problem anymore

mposa commented on systems/controllers/osc/osc_user_defined_traj.h:
```diff
@@ -0,0 +1,41 @@
+// TODO(yminchen):
```
If this isn't going to be included in the PR, best to remove it for now.

yminchen replied:
Yes, I'll move the reduced order model related code to my branch

yminchen commented on common/math_utils.h:
```diff
@@ -0,0 +1,12 @@
+
+namespace dairlib {
+
+// If the norm of quaternion is 0, set it to unit quaternion. Otherwise,
```
Should I fix the 0 quaternion problem now, or can I post an issue to fix it in the future?

mposa commented on systems/controllers/osc/operational_space_control.h:
```diff
@@ -0,0 +1,238 @@
+
+  // `state_` is the finite state machine state when the contact constraint is
+  // active. If `state_` is empty, then the constraint is always active.
+  std::vector<int> state_;
```
`state_` probably deserves a better name.`fsm_state_when_active_` is one suggestion, but you can probably find something better!

mposa commented on systems/controllers/osc/operational_space_control.h:
```diff
@@ -0,0 +1,238 @@
+namespace systems {
+namespace controllers {
+
+/// `OperationalSpaceControl` takes in desired trajectory in world frame and
```
This is very helpful!

mposa commented on systems/controllers/osc/operational_space_control.cc:
```diff
@@ -0,0 +1,602 @@
+      std::string name_w_spr = tree_w_spr_.get_position_name(j);
+      if (name_wo_spr.compare(0, name_wo_spr.size(), name_w_spr) == 0) {
+        map_position_from_spring_to_no_spring_(i, j) = 1;
+      }
```
throw an error if no match is found?

mposa commented on systems/controllers/osc/operational_space_control.cc:
```diff
@@ -0,0 +1,602 @@
+  }
+}
+
+void OperationalSpaceControl::ConstructOSC() {
```
This isn't a constructor, so maybe `Compile` rather than `ConstructOSC`? It should also have a `is_compiled_` flag that's set, with appropriate errors if (1) compile is called twice or (2) it's run without being compiled.

yminchen replied:
I was looking for the verb, but couldn't think of a good one. Do you think _compile_ is fine because it also has the meaning of translating our code to machine code?

mposa commented on systems/controllers/osc/operational_space_control.cc:
```diff
@@ -0,0 +1,602 @@
+        MatrixXd W = tracking_data->GetWeight();
+        MatrixXd J_t = tracking_data->GetJ();
+        VectorXd JdotV_t = tracking_data->GetJdotTimesV();
+        // Note that the following cost also include the constant term, so that
```
Helpful!

yminchen commented on systems/controllers/osc/osc_tracking_data.h:
```diff
@@ -0,0 +1,428 @@
+/// - gains of PD controller
+/// - cost weight
+/// - a flag indicating the trajectory is a constant
+/// - a flag indicating the trajecotry has exponential term (that is, the traj
```
Correct

yminchen replied:
I felt unsatisfied when I had to add an extra flag just because I couldn't use `AbstractValue<Trajectory> `, but it seems it's possible to initialize the abstract value now.

yminchen commented on systems/controllers/osc/osc_tracking_data.h:
```diff
@@ -0,0 +1,428 @@
+/// TransTaskSpaceTrackingData is used when we want to track a trajectory
+/// (translational position) in the task space.
+
+/// `track_center_of_mass_` should be set to true when we track center of mass.
```
I think it's probably better to have it within `TransTaskSpaceTrackingData`, because conceptually center of mass position lives in translational task space. It might be easier for the user to understand the whole `TrackingData` family if it stay inside `TransTaskSpaceTrackingData`? Additionally, `ComTrackingData`'d share methods `AddPointToTrack`

mposa commented:
Yes. Compile or Build are both pretty standard here, either is fine.

On Tue, Jul 23, 2019 at 9:49 PM Yu-Ming Chen <notifications@github.com>
wrote:

> *@yminchen* commented on this pull request.
> ------------------------------
>
> In systems/controllers/osc/operational_space_control.cc
> <https://github.com/DAIRLab/dairlib/pull/86#discussion_r306599047>:
>
> > +    DRAKE_DEMAND((W_joint_accel_.rows() == n_v_) &&
> +                 (W_joint_accel_.cols() == n_v_));
> +  }
> +}
> +void OperationalSpaceControl::CheckConstraintSettings() {
> +  if (!body_index_.empty()) {
> +    DRAKE_DEMAND(mu_ != -1);
> +    DRAKE_DEMAND(body_index_.size() == pt_on_body_.size());
> +  }
> +  if (!state_.empty()) {
> +    DRAKE_DEMAND(state_.size() == body_index_.size());
> +    DRAKE_DEMAND(state_.size() == pt_on_body_.size());
> +  }
> +}
> +
> +void OperationalSpaceControl::ConstructOSC() {
>
> I was looking for the verb, but couldn't think of a good one. Do you think
> *compile* is fine because it also has the meaning of translating our code
> to machine code?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/DAIRLab/dairlib/pull/86?email_source=notifications&email_token=ABBICGGU2EPX6C6BHS4NGZ3QA6YMLA5CNFSM4IF2QYCKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB7LSEXY#discussion_r306599047>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBICGCYVPISHSGXNM5O4KTQA6YMLANCNFSM4IF2QYCA>
> .
>

yminchen commented on systems/controllers/osc/osc_utils.cc:
```diff
@@ -0,0 +1,43 @@
+#include "systems/controllers/osc/osc_utils.h"
```
I moved the function under `OperationalSpaceControl` and made it a static member function

yminchen commented on systems/controllers/osc/osc_tracking_data.h:
```diff
@@ -0,0 +1,428 @@
+                               int body_index_wo_spr,
+                               Eigen::Vector3d pt_on_body = Eigen::Vector3d::Zero());
+
+  // (A version for std::vector in case the users want it)
```
I was not sure if I needed it. I'll delete them.

yminchen commented on systems/controllers/lipm_traj_gen.cc:
```diff
@@ -0,0 +1,212 @@
+  Y[1](0, 0) = stance_foot_pos(0);
+  Y[0](1, 0) = stance_foot_pos(1);
+  Y[1](1, 0) = stance_foot_pos(1);
+  Y[0](2, 0) = (CoM(2) > 0) ? CoM(2) : 1e-3;
```
It's a dead code which I will remove.

yminchen commented on systems/controllers/time_based_fsm.h:
```diff
@@ -0,0 +1,62 @@
+
+/// Constructor inputs:
+/// - rigid body tree
+/// - integer representing the first state
```
I'll use 0 and 1

yminchen commented on systems/controllers/time_based_fsm.h:
```diff
@@ -0,0 +1,62 @@
+
+/// There are three outcome of the states, beased on the current time t.
+/// If t < time_shift,
+///     state = `initial_state_number_`.
```
There was one time I needed to wait for the robot to fall on to the ground (touch the ground) in the beginning of simulation before I start the controller

yminchen commented on systems/controllers/time_based_fsm.cc:
```diff
@@ -0,0 +1,68 @@
+    const Context<double>& context,
+    BasicVector<double>* fsm_state) const {
+  // Read in current state and simulation time
+  const OutputVector<double>* robot_output = (OutputVector<double>*)
```
I use robot_output to get the current time, but I can also get the time from context, so I will delete the port.

yminchen replied:
I noticed that the robot walks better when it's using lcm time, because context time is slower than lcm time. I think I'll retain the state port for now.

mposa replied:
Leave for now, likely related to https://github.com/DAIRLab/dairlib/issues/63

yminchen commented:
@mposa I've made changes based on your feedback except TrajectoryWrapper so far.

yminchen commented:
@mposa I'm done with revising codes.

mposa commented on systems/controllers/osc/osc_tracking_data.cc:
```diff
@@ -0,0 +1,387 @@
+          K_p, K_d, W, tree_w_spr, tree_wo_spr) {
+}
+
+void RotTaskSpaceTrackingData::AddFrameToTrack(std::string body_name,
```
This comment is for this TD class and others.

Won't everything get out of sync, if you were to call the sequence below?
```
AddFrameToTrack(name1, pose1);
AddStateAndFrameToTrack(state, name2, pose2);
```
I don't think this code will properly assign the first frame to all FSM states, and the second to only one.

yminchen replied:
At most one point (of the body) can follow the desired trajectory, so `state_` elements can not repeat, and the length of `state_` must be the same as `frame_pose_`'s if `state_` is not empty.

There is a method `CheckDerivedOscTrackingData` that checks if the `state_` length is correct.

I just added comments about this, so it's more obvious for the users.

mposa replied:
Ok for now, but will add an issue.

mposa commented on systems/controllers/osc/operational_space_control.h:
```diff
@@ -0,0 +1,246 @@
+  }
+
+  // Osc leafsystem builder
+  void BuildOSC();
```
Why not just `Build`? The class is `OSC`, so it's unnecessary to specify that twice.

yminchen replied:
Right, Build sounds better

mposa commented on systems/controllers/osc/operational_space_control.cc:
```diff
@@ -0,0 +1,619 @@
+  n_q_ = tree_wo_spr.get_num_positions();
+  n_v_ = tree_wo_spr.get_num_velocities();
+  n_u_ = tree_wo_spr.get_num_actuators();
+  cout << "n_q_ = " << n_q_ << endl;
```
Can these cout lines be safely removed?

mposa commented on systems/controllers/osc/operational_space_control.cc:
```diff
@@ -0,0 +1,619 @@
+}
+
+// Constraint methods
+void OperationalSpaceControl::AddContactPoint(std::string body_name,
```
Same comment here as for the tracking data. Won't this all get out of sync if both `AddContactPoint` and `AddStateAndContactPoint` are called? A better way of tracking "all states" is needed than emptiness of the state vector.

yminchen replied:
There is a private method `CheckConstraintSettings` that checks if the state length is correct.

mposa commented on systems/controllers/osc/operational_space_control.cc:
```diff
@@ -0,0 +1,619 @@
+    VectorXd mu_minus1(2); mu_minus1 << mu_, -1;
+    VectorXd mu_plus1(2); mu_plus1 << mu_, 1;
+    VectorXd inf_vectorxd(1); inf_vectorxd << numeric_limits<double>::infinity();
+    for (unsigned int i = 0; i < active_contact_flags.size(); i++) {
```
Probably `inf`--but better yet would be an efficient way to just remove the constraint.

yminchen replied:
I can't find a way to remove a constraint. 
I wonder if the total runtime is short if we remove and add the constraint compared to only updating the coefficients?

mposa replied:
Leave it as a TODO for now.

yminchen commented on attic/multibody/rigidbody_utils.h:
```diff
@@ -31,5 +31,17 @@ int GetBodyIndexFromName(const RigidBodyTree<double>& tree, std::string name);
+/// whatever is causing the 0 quaternion state in the controller.
+///
+/// Input: the configuration of the robot which must be in floating-base
+void SetZeroQuaternionToIdentity(Eigen::VectorXd* q);
```
Do I need to write more comment? There is already one line describing the input.

mposa replied:
Just saying that it's in floating base isn't quite clear enough. This code very explicitly expects the quaternion to be elements 3-6.

yminchen commented on systems/controllers/control_utils.h:
```diff
@@ -0,0 +1,53 @@
+///
+/// Requirement:
+///  - The controller is used with two-state finite state machine
+Eigen::Vector2d ImposeHalfplaneGaurd(Eigen::Vector2d foot_placement_pos,
```
Actually this can be used more broadly. The only requirement for using this function is that the robot has to be bipedal. 
I'll replace `right/left stance state` and `state` with one boolean `left_right_stance_flag`.

mposa replied:
It's OK, but feels pretty specific. Fine to leave for now.

yminchen commented:
@mposa I've pushed the changes!

mposa commented:
Pending that last comment re:documenting quaternion indices, ready to go. Please provide a summary commit message to use for the entire PR.

yminchen commented:
@mposa Just finished the last requested update.
Here is the summary of all the changes:

Major updates:
- Added `OperationalSpaceControl` and the related classes
- Added standing controller for Cassie based on operational space control
- Added walking controller for Cassie based on operational space control and capture point foot placement

Minor updates:
- Added time based finite state machine
- Updated `cassie_utils` to create fixed spring cassie model
- Updated `cassie_fixed_springs.urdf` so that it's the same as `cassie_v2.urdf` except the springs
- Updated `cassie_v2.urdf` to include toe motor limit 
- Added a util function to check if a RBT is floating based
- Added flags to initialize Cassie's pose in `run_simple_sim.cc`
- Added a function to set 0-norm quaternion to identity
